browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,461 @@
+
+  base::Optional<base::Value> regional_selectors = g_brave_browser_process->
+      ad_block_regional_service_manager()->
+          HiddenClassIdSelectors(classes, ids, exceptions);
+
+  base::Optional<base::Value> custom_selectors = g_brave_browser_process->
+      ad_block_custom_filters_service()->
+          HiddenClassIdSelectors(classes, ids, exceptions);
+
+  if (hide_selectors && hide_selectors->is_list()) {
+    if (regional_selectors && regional_selectors->is_list()) {
+      for (auto i = regional_selectors->GetList().begin();
+          i < regional_selectors->GetList().end(); i++) {
+        hide_selectors->Append(std::move(*i));
+      }
+    }
+  } else {
+    hide_selectors = std::move(regional_selectors);
+  }
+
+  auto result_list = std::make_unique<base::ListValue>();
+  if (hide_selectors && hide_selectors->is_list()) {
+    result_list->Append(std::move(*hide_selectors));
+  }
+  if (custom_selectors && custom_selectors->is_list()) {
+    result_list->Append(std::move(*custom_selectors));
+  }
+
+  return result_list;
+}
+
+bool IsVettedSearchEngine(const std::string& host) {
+  for (size_t i = 0;
+      i < BraveCosmeticResourcesTabHelper::vetted_search_engines_.size();
+      i++) {
+    size_t found_pos = host.find(
+        BraveCosmeticResourcesTabHelper::vetted_search_engines_[i]);
+    if (found_pos != std::string::npos) {
+      size_t last_dot_pos = host.find(".", found_pos + 1);
+      if (last_dot_pos == std::string::npos) {
+        return false;
+      }
+      if (host.find(".", last_dot_pos + 1) == std::string::npos) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+}  // namespace
+
+
+BraveCosmeticResourcesTabHelper::BraveCosmeticResourcesTabHelper(
+    content::WebContents* contents)
+    : WebContentsObserver(contents),
+    enabled_1st_party_cf_filtering_(false) {
+  if (BraveCosmeticResourcesTabHelper::observing_script_->empty()) {
+    *BraveCosmeticResourcesTabHelper::observing_script_ =
+        LoadDataResource(kCosmeticFiltersGenerated[0].value);
+  }
+}
+
+BraveCosmeticResourcesTabHelper::~BraveCosmeticResourcesTabHelper() {
+}
+
+void BraveCosmeticResourcesTabHelper::GetUrlCosmeticResourcesOnUI(
+    content::GlobalFrameRoutingId frame_id, const std::string& url,
+    bool do_non_scriplets, std::unique_ptr<base::ListValue> resources) {
+  if (!resources) {
+    return;
+  }
+  for (auto i = resources->GetList().begin();
+      i < resources->GetList().end(); i++) {
+    base::DictionaryValue* resources_dict;
+    if (!i->GetAsDictionary(&resources_dict)) {
+      continue;
+    }
+    std::string to_inject;
+    resources_dict->GetString("injected_script", &to_inject);
+    if (to_inject.length() > 1) {
+      auto* frame_host = content::RenderFrameHost::FromID(frame_id);
+      if (!frame_host)
+        return;
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+          base::UTF8ToUTF16(to_inject),
+          base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
```

Comment:
this will make the injected script run in the same js context as some other chromium features: https://source.chromium.org/search?q=ISOLATED_WORLD_ID_CHROME_INTERNAL&sq=&ss=chromium

seems slightly safer to run it in its own world unless all the injected scripts are carefully audited

Reply:
All the scriptlets come from our uBlock Origin fork at https://github.com/brave/uBlock, which is manually reviewed before updates are applied

Reply:
@diracdeltas I did diff ID initially only for us, but @bridiver corrected me, I will probably let him comment on that, his words were something like:
we will be using the same id with chrome if we use ISOLATED_WORLD_ID_CHROME_INTERNAL, but that doesn't matter
the only reason that they use multiple IDs for extensions is to prevent them from accessing each other
but that's not an issue with our own internal code

Reply:
@pes10k kindly confirmed that scriptlet updates are manually audited but cosmetic filters aren't

Reply:
@diracdeltas @SergeyZhukovsky i think this looks good, though a have a small nit request.  When you're creating JS strings to inject from C++ (or other languages) could you construct the string all in one place (or better yet, read it from disk in a separate file) with some %% style token, and then std::string::replace (or similar) to just drop in the values you're inserting?  It'd make it easier to confidently understand the final JS code (and if it was in a separate file, you could even use static analyzers etc.

Not a substantive thing, but something that'd make it to be certain there is nothing unexpected happening.

But on the substance, this looks right to me!